contracts/cw-names-registry/src/contract.rs

```diff
@@ -0,0 +1,256 @@
+
+    config.payment_amount = payment_amount;
+    config.admin = admin;
+    config.payment_token_address = payment_token_address;
+
+    CONFIG.save(deps.storage, &config)?;
+
+    Ok(Response::new().add_attribute("action", "update_config"))
+}
+
+pub fn execute_revoke(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    name: String,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    let dao = NAME_TO_DAO.load(deps.storage, name.clone())?;
+
+    // Only name owner and overall name admin can revoke
+    if info.sender != dao && info.sender != config.admin {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    NAME_TO_DAO.remove(deps.storage, name);
+    DAO_TO_NAME.remove(deps.storage, dao);
+
+    Ok(Response::new())
+}
+
+pub fn execute_reserve(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    name: String,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    // Check if name is already taken
+    if NAME_TO_DAO.has(deps.storage, name.clone()) {
+        return Err(ContractError::NameAlreadyTaken {});
+    }
+
+    // Only the admin can reserve names
+    if info.sender != config.admin {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    RESERVED_NAMES.save(deps.storage, name, &Empty {})?;
+
+    Ok(Response::new())
+}
+
+pub fn execute_transfer_reservation(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    name: String,
+    dao: String,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    // Validate DAO
+    let dao = deps.api.addr_validate(&dao)?;
+
+    // Only the admin can transfer reserved names
+    if info.sender != config.admin {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    // Check if DAO already has a name
+    if DAO_TO_NAME.has(deps.storage, dao.clone()) {
+        return Err(ContractError::NameAlreadyRegistered {});
+    }
+
+    if !RESERVED_NAMES.has(deps.storage, name.clone()) {
+        return Err(ContractError::NameNotReserved {});
+    }
+
+    DAO_TO_NAME.save(deps.storage, dao.clone(), &name)?;
+    NAME_TO_DAO.save(deps.storage, name.clone(), &dao)?;
+    RESERVED_NAMES.remove(deps.storage, name);
+
+    Ok(Response::new())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
+        QueryMsg::Config {} => to_binary(&CONFIG.load(deps.storage)?),
+        QueryMsg::LookUpDao { dao } => query_look_up_dao(deps, dao),
+        QueryMsg::LookUpName { name } => query_look_up_name(deps, name),
+    }
+}
+
+pub fn query_look_up_dao(deps: Deps, dao: String) -> StdResult<Binary> {
+    let dao = deps.api.addr_validate(&dao)?;
+    let name = DAO_TO_NAME.may_load(deps.storage, dao)?;
+    to_binary(&LookUpDaoResponse { name })
+}
+
+pub fn query_look_up_name(deps: Deps, name: String) -> StdResult<Binary> {
```

Comment:
I had to read this method a few times, and I'm still just guessing at what it does. That might be my lack of having my head in contracts. But I think a different name could go some way here.

My guess is that this contract tells you (1) whether a name is available or not and (2) if it is unavailable, to what DAO it refers.

Does this method give you a binary (whether or not this name is available)? If so, a more descriptive name might be `look_up_dao_by_name` (and `LookUpDaoByName`).

Similarly, `LookUpNameByDao`, `query_look_up_name_by_dao`.